### PR TITLE
feat(scheduler): share compute and interleave bounded prefills

### PR DESCRIPTION
**Pull request by voipmonitor**
## Purpose

Keep active decodes responsive while long prefills make bounded progress.
The scheduler targets a prefill share of **engine-observed model service
time** and can interleave several prefill requests. This is wall-clock
completion feedback, not CUDA-event measurement of pure GPU kernel time.

## Behavior

- `--prefill-compute-share FLOAT|auto` enables feedback-based sharing. The
  vLLM default remains disabled; the GLM community launcher selects fixed 0.4.
- `--prefill-compute-half-life` controls automatic-share smoothing.
- `--max-parallel-prefills`, `--prefill-policy`, and
  `--decode-refill-target` expose bounded round-robin/decode-aware admission.
  One lane remains the default. `auto` resolves to at most four lanes,
  independently of attention pages, recurrent checkpoints, or LMCache objects.
  The global token budget and request priority remain authoritative.
- Only contended model work requests timing. A primitive timestamp stays
  paired with each executor batch; completion accounting excludes already
  charged queue residency and attaches no timing callback to the future.
- Decode-only execution bypasses redundant runnable-decode and prefill-lane
  scans. Automatic demand observation remains separate; this does not make
  the entire scheduler constant-time.
- External cache restores consume neither local-prefill lanes nor compute
  credit while waiting for transfer. A recurrent boundary-logits restore owns
  one complete, isolated scheduler step.

Derek Yates's compute-share, interleaving, and hot-path hardening contributions
retain authorship. Boundary isolation and the GLM integration tests preserve
the request-checkpoint contract. This PR is the canonical review target for
the work from closed #648; do not stack that integration-based PR separately.
This updates the existing scheduler PR rather than creating a duplicate.
Implementation and validation used AI assistance (Codex); commit attribution
identifies the human contributors and the assistant.

## JJ composition contract

Status: **implemented; focused CPU contracts qualified**. The branch includes JJ's model additions. This PR owns the connector readiness hook and releases prefill lanes while atomic external checkpoint imports wait. Merge this PR and #553 before #709; #709 supplies the atomic transfer implementation. Both pending-import cases (one and two prefill lanes) are tested without consuming recurrent state. The branch passes 117 focused CPU tests; the composed scheduler/checkpoint stack passes 398. These counts do not substitute for GPU serving qualification of added model dependencies.

## Recorded R28.1 validation

**Qualified:** 156 focused CPU tests on this PR branch; 159 focused tests on
the complete FP8 serving integration, repeated against its installed image.
Coverage includes disabled timing, exact batch/timestamp pairing, automatic
lanes at a 4096-token budget with 4096-token storage geometry, priority,
asynchronous transfers, and recurrent boundary isolation across share/policy
settings. Ruff and `git diff --check` pass.

Focused suite invocation in the image's isolated Python environment:

```bash
VLLM_TARGET_DEVICE=cpu PYTORCH_CUDA_ALLOC_CONF= \
  /opt/venv/bin/python -m pytest -q \
  tests/v1/core/test_compute_fairness.py \
  tests/v1/core/test_prefill_compute_share_scheduler.py \
  tests/v1/engine/test_compute_fairness_feedback.py \
  tests/v1/engine/test_prefill_fairness_runtime.py \
  tests/entrypoints/serve/dev/test_fairness.py
```

This five-suite installed-image invocation passes 135 tests. The test-only
`tblib` dependency and the CLI's cached default model metadata are supplied
read-only outside the installed serving package; no model executes.
GPU-free CLI tests select the CPU platform explicitly; mocked KV
connectors do not inherit CUDA expandable-segment settings. Neither test
environment adjustment changes the qualified GPU serving configuration.

The two-slot engine reproducer also covers an untimed predecessor completing
after a timed successor was dispatched: its queue residency is excluded
without enabling timing for wholly uncontended queues. The merge with JJ
preserves reserved boundary-logits admission fallback when cache allocation,
LoRA limits, input readiness, or request policy blocks the isolated request.

**Qualified FP8 serving and checkpoints:** exact R28.1 versus R28 on the same
stock TP4 RTX PRO 6000 Workstation quartet, budget4096, OMP1, NCCL16/2MiB,
full-and-piecewise graphs and explicit max reasoning on both images.

- MTP3/DCP4 C64, three 60-second cells: 2153.66 → 2155.56 output tok/s
  (+0.09%), 844.67 → 846.09 verifier steps/s (+0.17%).
- 32K prefill: MTP3/DCP4 −0.19%, no-spec/DCP1 +0.04%, DFlash2/DCP1 +0.02%.
- Four-lane MTP3/DCP4 passes 1M RAM and restart-filesystem restore with zero
  recompute, shared SYSTEM reuse, literal answers, all-rank byte integrity and
  C8 cancellation/read-eviction. DFlash aligned-256 interior-prefix checks pass.
- With four active decodes and eight long prefills, late4K TTFT changes
  46.85 → 8.73 s for one → four lanes; median long-request TTFT increases
  approximately 26.1 → 47.3 s. All requests complete. This is a latency
  trade-off; one lane remains the default.

**Performance limitation:** DFlash C8's short cell is −2.39% output / −0.98%
verifier. Extended repeats were not run; no blanket equivalence or speedup is
claimed. MTP's retained short C1 dip did not persist in three longer repeats.
The [complete report and raw summaries](https://github.com/local-inference-lab/rtx6kpro/blob/master/models/glm-5.3-flash/validation/scheduler-serving-r28.1.md)
record those observations, artifact identities and qualification boundaries.
Pipeline parallel completion accounting, TP8 and NVFP4 target KV are not
qualified here. B12X, FlashKDA and native libraries are byte-identical to R28.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

**Comment by voipmonitor** [comment]
Status: active; this is the replacement for PR #648.

dev/jovian-judgement at db7a65e295 does not contain the bounded multi-prefill interleaver or its measured compute-share controller. The target branch also gained recurrent request-boundary checkpoint behavior after this branch was prepared. The logits-only restore step must remain isolated as one complete scheduler step after rebasing; it must not be co-scheduled with prefill or decode work.

The present branch conflicts with the target. Rebase it onto the target head, preserve Derek Yates authorship for the scheduler commits, and repeat scheduler tests, exact-prefix restore tests, and mixed prefill/decode qualification. PR #648 must remain closed and must not be cherry-picked separately.

**Comment by voipmonitor** [comment]
## R27 integration validation

The change represented by this PR is included in the qualified, source-locked
GLM-5.3-Flash runtime
`voipmonitor/vllm:jovian-judgement-community-20260906-r27`
(`sha256:a298fe1cd207eaf97bd2ff2686716ed25b7009c09b36650eba732a4a7dc51512`).
The exact vLLM composition is mirrored at
[`voipmonitor/vllm:integration/glm53-r27-release-20260906`](https://github.com/voipmonitor/vllm/tree/integration/glm53-r27-release-20260906),
commit `63a82f8d323e8538cbe6f88ae1812a1c01577a0f`.

Qualification used four stock-clock RTX PRO 6000 Blackwell Workstation Edition
GPUs, TP4, a 4,096-token scheduler budget, 16 NCCL channels, a 2 MiB NCCL
buffer, and full plus piecewise CUDA graphs:

| Mode | DCP | 32K prefill | C1 output / steps | C8 output / steps |
|---|---:|---:|---:|---:|
| No speculation | 1 | 14,870 tok/s | 170.6 tok/s | 733.8 tok/s |
| MTP3 | 1 | 14,468 tok/s | 276.0 / 109.1 tok/s | 901.0 / 371.3 tok/s |
| MTP3 full CKV | 4 | 12,864 tok/s | 247.0 / 97.1 tok/s | 876.4 / 346.4 tok/s |
| DFlash2 K7 full CKV, NVFP4 KV | 4 | 12,633 tok/s | 198.0 / 81.2 tok/s | 645.5 / 260.8 tok/s |

FP8 no-speculation and NVFP4 DFlash2 external-cache configurations also passed
cold compute, vLLM prefix reuse, engine-driven RAM-L1 restore, full-process
filesystem-L2 restore, and block-checksum validation on all four ranks. An exact
81,576-token leading-instruction test reused 81,567 tokens when only the user
continuation changed.

This is an integration and regression gate, not an isolated attribution of the
aggregate throughput to this PR. The complete open-PR merge order and evidence
are recorded in #651.

**Comment by voipmonitor** [comment]
The compute-share scheduler branch now merges cleanly with `dev/jovian-judgement` at `7d66922a7bf`. Derek's source attribution is preserved.

Validation on the merged branch:

- 261 scheduler, fairness-feedback, and prefix-cache CPU tests passed.
- All configured pre-commit checks passed, including mypy.
- Scheduler behavior matches the qualified R27 source. The additional scheduler edit is a variable-length tuple annotation; it does not change scheduling.

This merge resolution preserves the independent logits-only checkpoint step and compute-share fairness. It does not reintroduce the removed micro-slicing controller.

@coderabbitai review

**Comment by voipmonitor** [comment]
The scheduler hot-path hardening from Derek Yates's #648 is now included here,
with authorship preserved:

- Engine-owned primitive timestamps replace the timing helper and its future
  callback. Only contended model service is timed.
- Decode-only iterations bypass redundant runnable-decode and prefill-admission
  scans. Automatic demand observation remains separate.
- Automatic lanes resolve to `min(4, max_num_seqs)` regardless of cache geometry;
  explicit lane counts are no longer capped by `4096 / 4096` either.

Validation: **127 focused CPU tests on this PR branch**, plus **151 tests on the
complete R28-based serving composition and its installed image**. The latter
include boundary-logits isolation under fixed/auto share and both admission
policies. The hardening patch is identical in the PR and image composition.

The launcher mapping and authenticated Python-only overlay are in
blackwell-llm-docker#31. The image retains share 0.4 and one lane by default.
Same-quartet TP4/DCP4 FP8 MTP3 A/B, mixed-workload, and cache checks are running;
GPU conclusions will be appended before release. No speedup is claimed from
the CPU checks.

@coderabbitai review

**Comment by voipmonitor** [comment]
The queue-residency finding is reproducible, including TP-only asynchronous
scheduling with its two executor slots. It is not specific to automatic share.

A regression test uses the real `step_with_batch_queue()` loop: dispatch the
contended successor at 10.1 s, observe its untimed predecessor complete at
10.2 s, then observe the successor at 10.5 s. The previous accounting charged
0.4 s instead of 0.3 s. Both a transfer-only predecessor and an uncontended
model predecessor reproduced it.

The fix is in this PR. After consuming an untimed result, the engine records
its completion boundary **only if its already-dispatched successor needs
timing**. It covers queued sampling and deferred grammar sampling. The
predecessor receives no compute charge; a wholly uncontended queue still
performs zero clock reads. No future callbacks or always-on transfer timers
are introduced.

Validation: eight engine-loop regressions pass, covering transfer/model
predecessors, timed/untimed successors, and deferred sampling. All 135 focused
tests pass on this PR; all 159 pass on the serving integration. Mypy, Ruff, and
the repository commit checks pass. The serving image is being rebuilt and its
GPU regression checks remain pending.

@coderabbitai review

**Comment by voipmonitor** [comment]
The scheduler composition is published in
`voipmonitor/vllm:jovian-judgement-community-20260908-r28.1`.
The PR body now includes the CPU commands, queued-completion reproducer,
same-quartet GPU results and performance limits.

MTP3/DCP4 C64: 2153.66 → 2155.56 tok/s (+0.09%),
844.67 → 846.09 verifier steps/s (+0.17%). All three mode controls keep 32K
prefill within 0.2%. Four-lane RAM/restart checkpoint, cancellation, priority
and aligned-DFlash checks pass. Short DFlash C8 is −2.39% output / −0.98%
verifier; extended repeats were not run, and the observation remains visible.

Four lanes reduce late4K TTFT under a long-prefill convoy from 46.85 to 8.73 s,
but long-request median TTFT increases from about 26.1 to 47.3 s. One lane
therefore remains the image default. This is not a blanket throughput claim.

The complete serving mirror retains the #709 external-import lane-credit
resolution and the JJ reserved boundary-logits admission fallback. Native
components are unchanged. Derek Yates's hardening attribution is preserved.

[Full qualification and raw samples](https://github.com/local-inference-lab/rtx6kpro/blob/master/models/glm-5.3-flash/validation/scheduler-serving-r28.1.md).

@coderabbitai review

**Comment by voipmonitor** [comment]
Review composition validation: this PR owns readiness polling and releases prefill lanes while external imports are pending; #709 owns atomic transfer. Final composed JJ passes 393 focused CPU tests, including pending imports and timing/admission contracts. Packaged TP4/DCP1 DFlash C1/C8/32K and prefix checks pass within the bounded report. Merge this PR and #553 before #709.

Evidence and limitations: https://github.com/local-inference-lab/blackwell-llm-docker/blob/codex/glm53-source-locked-build/recipes/glm53/review-qualification.md

Validation performed with OpenAI Codex assistance.

@coderabbitai review

**Comment by voipmonitor** [comment]
Included in `dev/jovian-judgement` through [this PR's individual merge](https://github.com/local-inference-lab/vllm/commit/31ab94e2a34e77f25758de941e2d614bceed98f3). The reviewed head and its contributor commits remain ancestors; the PR is merged and closed.

Source validation: replaying all 32 R35 review heads on the pinned base exactly reproduces the released Docker's vLLM tree; all 6,870 installed tracked files match. JJ additionally preserves Luke's DS4.1 work and #734. The final composition passed 247 focused checkpoint/scheduler, sampler/warmup and native GPU tests. This is combined-source evidence, not a fresh performance or full-model qualification for this individual PR.

Publication-history clarification: the individual merge linked above is in JJ's first-parent history. It replaces the receipt's archived wrapper-merge reference; GitHub's historical merge SHA may still identify that archive. See #731 for component review order and qualification limits.
